writers/ScatterGatherBuffer.go

```diff
@@ -137,15 +138,40 @@ func (sgb *ScatterGatherBuffer) IsEmpty() bool {
 	return sgb.size == 0
 }
 
-func (sgb *ScatterGatherBuffer) ToBytes() []byte {
-	result := make([]byte, sgb.size)
-	offset := 0
+func (sgb *ScatterGatherBuffer) GetPipeReader() *io.PipeReader {
+	reader, writer := io.Pipe()
 
-	for i := 0; i < sgb.buffer.Len(); i++ {
-		chunk := sgb.buffer.At(i)
-		copy(result[offset:], chunk)
-		offset += len(chunk)
+	go func() {
+		defer writer.Close()
+
+		for i := 0; i < sgb.buffer.Len(); i++ {
+			chunk := sgb.buffer.At(i)
+			_, err := writer.Write(chunk)
+			if err != nil {
+				_ = writer.CloseWithError(err)
+				return
+			}
+		}
+	}()
```

Comment:
This is a fine implementation for now, but it makes much more sense to have `struct ScatterGatherBuffer` implement `io.Reader` directly, and have `Read()` copy `n` bytes to the output buffer and drop them from the sgb. This way you are going to have many fewer grouting wakeups and switches.